flytekit/models/literals.py

```diff
@@ -979,7 +979,15 @@ def offloaded_metadata(self) -> Optional[LiteralOffloadedMetadata]:
         """
         This value holds metadata about the offloaded literal.
         """
-        return self._offloaded_metadata
+        # The following check might seem non-sensical, since `_offloaded_metadata` is set in the constructor.
+        # This is here to support backwards compatibility caused by the local cache implementation. Let me explain.
+        # The local cache pickles values and unpickles them. When unpickling, the constructor is not called, so there
+        # are cases where the `_offloaded_metadata` is not set (for example if you cache a value using flytekit<=1.13.6
+        # and you load that value later using flytekit>1.13.6).
+        # In other words, this is a workaround to support backwards compatibility with the local cache.
+        if hasattr(self, "_offloaded_metadata"):
```

Comment:
> if isinstance(serialized_obj, ModelLiteralMap):
            return ModelLiteralMap.from_flyte_idl(ModelLiteralMap.to_flyte_idl(serialized_obj))

Q: we double deserialize the cached value. Won't it have `_offloaded_metadata` in the literal map?

Reply:
We need this because `ModelLiteralMap.to_flyte_idl` uses the `offloaded_metadata` property [directly](https://github.com/flyteorg/flytekit/blob/cc9f96171a08339b9f5f63e89ddb2079f18fab8e/flytekit/models/literals.py#L1002). The idea is that by forcing it to go through `ModelLiteralMap.from_flyte_idl` we invoke the constructor which sets the [`_offloaded_metadata` field](https://github.com/flyteorg/flytekit/blob/master/flytekit/models/literals.py#L1013-L1015) (which bulletproofs the literal model file from similar changes going forward).

Reply:
I see, thank you